lua/entities/gmod_wire_datasocket.lua

```diff
@@ -8,9 +8,9 @@ function ENT:GetPlugClass()
 end
 
 if CLIENT then
-	hook.Add("HUDPaint","Wire_DataSocket_DrawLinkHelperLine",function()
-		local sockets = ents.FindByClass("gmod_wire_datasocket")
-		for k,self in pairs( sockets ) do
+	local sockets = ents.FindByClass("gmod_wire_datasocket") or {}
```

Comment:
I thought ENT:Initialize would be called when the player loads in for each socket. If this is the case then it shouldn't matter if they're connected to the server when the entity is placed or not. I'll do a test real quick just to make sure.

Reply:
> I thought ENT:Initialize would be called when the player loads in for each socket. If this is the case then it shouldn't matter if they're connected to the server when the entity is placed or not. I'll do a test real quick just to make sure.

the problem is you're setting `sockets = findByClass(whatever)` *but also* using `table.insert` in intialize which means the length of the array will never be 1 if someone already has one spawned in. it'll jump from 0 to 2 immediately.

that is, if the initialize function is called for each entity. which it probably is like I said.

Reply:
> unless, for some strange reason, `findByClass` doesn't work when used that early in the loading process, meaning it always returns empty array. in which case you might as well erase it and just initialize the variable to an empty array always.

I believe this is the case since the findByClass called is before InitPostEntity. I added the findByClass call for lua refresh compatibility like you suggested previously since a lua refresh in that file would break the hooks by removing the entities from the local table. I'm pretty sure the code will work without changes, but I will test.

Reply:
I spawned in 3 sockets before I loaded in to test. The code appears to work fine as is because of the fact ents.FindByClass is called so early that it will only ever return an actual table of entities if the file is lua refreshed.

Changed code to this for testing:

```lua
	local sockets = ents.FindByClass("gmod_wire_socket") or {}
	print("printing entities")
	PrintTable(ents.GetAll())
	print([[printing sockets:]])
	PrintTable(sockets)
```

```lua
	function ENT:Initialize()
		table.insert(sockets, self)
		if #sockets == 1 then
			print("Wire_Socket_DrawLinkHelperLine hook added")
			hook.Add("HUDPaint", "Wire_Socket_DrawLinkHelperLine",DrawLinkHelperLinefunction)
		end
	end
```

Output: 

```
printing entities
printing sockets:
```

```
Wire_Socket_DrawLinkHelperLine hook added
```